File: src/api/fileVideo.ts
import { post } from '../util/http'
import { normalizeFileCdnUrl } from '@/util/qiniuCdn'

const PLAY_TIMEOUT_MS = 60000

export type VideoPlayMode = 'mp4' | 'hls' | 'processing' | 'unsupported'

export type PlayFileVideoResp = {
  code: number
  message: string
  mode: VideoPlayMode
  url?: string
  playlist?: string
  duration_ms: number
}

export type FileVideoMarker = {
  id: number
  file_id: number
  time_ms: number
  label: string
}

type MarkerListResp = {
  code: number
  message: string
  markers: FileVideoMarker[]
}

type MarkerOkResp = {
  code: number
  message: string
}

function normalizePlaylist(text: string): string {
  return text.replace(/https?:\/\/cloud\.onij\.fun\/[^\s"']+/gi, (u) => normalizeFileCdnUrl(u))
}

export const PlayFileVideo = async (fileId: number): Promise<PlayFileVideoResp> => {
  const resp = (await post(
    '/file/video/play',
    { file_id: fileId },
    { timeout: PLAY_TIMEOUT_MS },
  )) as PlayFileVideoResp
  if (resp.url) resp.url = normalizeFileCdnUrl(resp.url)
  if (resp.playlist) resp.playlist = normalizePlaylist(resp.playlist)
  return resp
}

export const ListFileVideoMarkers = async (fileId: number): Promise<MarkerListResp> => {
  const resp = (await post('/file/video/marker/list', { file_id: fileId })) as MarkerListResp
  resp.markers = resp.markers ?? []
  return resp
}

export const SaveFileVideoMarkers = async (
  fileId: number,
  markers: Array<Pick<FileVideoMarker, 'time_ms' | 'label'> & { id?: number }>,
): Promise<MarkerListResp> => {
  const resp = (await post('/file/video/marker/save', {
    file_id: fileId,
    markers,
  })) as MarkerListResp
  resp.markers = resp.markers ?? []
  return resp
}

export const DeleteFileVideoMarker = async (id: number): Promise<MarkerOkResp> => {
  return post('/file/video/marker/delete', { id }) as Promise<MarkerOkResp>
}

File: server/biz/prm/file_video.go
package prm

import (
	"onij/model"
	"onij/model/api"
	"onij/util"
)

const (
	VideoPlayModeMP4         = "mp4"
	VideoPlayModeHLS         = "hls"
	VideoPlayModeProcessing  = "processing"
	VideoPlayModeUnsupported = "unsupported"
)

type PlayFileVideoParam struct {
	FileId int64
}

func NewPlayFileVideoParam(req *api.PlayFileVideoReq) *PlayFileVideoParam {
	return &PlayFileVideoParam{FileId: req.FileId}
}

type PlayFileVideoResult struct {
	Mode       string
	URL        string
	Playlist   string
	DurationMs int64
	Hint       string
}

func (r *PlayFileVideoResult) Resp() *api.PlayFileVideoResp {
	msg := util.BaseMsgOK
	if r.Hint != "" {
		msg = r.Hint
	}
	return &api.PlayFileVideoResp{
		Code:       util.BaseCodeOK,
		Message:    msg,
		Mode:       r.Mode,
		Url:        r.URL,
		Playlist:   r.Playlist,
		DurationMs: r.DurationMs,
	}
}

func ToFileVideoMarkerDTO(m *model.FileVideoMarker) *api.FileVideoMarkerItem {
	if m == nil {
		return nil
	}
	return &api.FileVideoMarkerItem{
		Id:     m.Id,
		FileId: m.FileId,
		TimeMs: m.TimeMs,
		Label:  m.Label,
	}
}

type ListFileVideoMarkerParam struct {
	FileId int64
}

func NewListFileVideoMarkerParam(req *api.ListFileVideoMarkerReq) *ListFileVideoMarkerParam {
	return &ListFileVideoMarkerParam{FileId: req.FileId}
}

type ListFileVideoMarkerResult struct {
	Markers []*api.FileVideoMarkerItem
}

func (r *ListFileVideoMarkerResult) Resp() *api.ListFileVideoMarkerResp {
	markers := r.Markers
	if markers == nil {
		markers = []*api.FileVideoMarkerItem{}
	}
	return &api.ListFileVideoMarkerResp{
		Code:    util.BaseCodeOK,
		Message: util.BaseMsgOK,
		Markers: markers,
	}
}

type SaveFileVideoMarkerParam struct {
	FileId  int64
	Markers []*api.FileVideoMarkerItem
}

func NewSaveFileVideoMarkerParam(req *api.SaveFileVideoMarkerReq) *SaveFileVideoMarkerParam {
	return &SaveFileVideoMarkerParam{FileId: req.FileId, Markers: req.Markers}
}

type SaveFileVideoMarkerResult struct {
	Markers []*api.FileVideoMarkerItem
}

func (r *SaveFileVideoMarkerResult) Resp() *api.SaveFileVideoMarkerResp {
	markers := r.Markers
	if markers == nil {
		markers = []*api.FileVideoMarkerItem{}
	}
	return &api.SaveFileVideoMarkerResp{
		Code:    util.BaseCodeOK,
		Message: util.BaseMsgOK,
		Markers: markers,
	}
}

type DeleteFileVideoMarkerParam struct {
	Id int64
}

func NewDeleteFileVideoMarkerParam(req *api.DeleteFileVideoMarkerReq) *DeleteFileVideoMarkerParam {
	return &DeleteFileVideoMarkerParam{Id: req.Id}
}

type DeleteFileVideoMarkerResult struct{}

func (r *DeleteFileVideoMarkerResult) Resp() *api.DeleteFileVideoMarkerResp {
	return &api.DeleteFileVideoMarkerResp{
		Code:    util.BaseCodeOK,
		Message: util.BaseMsgOK,
	}
}
